src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -111,71 +91,42 @@ private abstract class NonFinalWithCompilationTrackerState : CompilationTrackerS
                 /// </summary>
                 public Compilation? StaleCompilationWithGeneratedDocuments { get; }
 
-                protected NonFinalWithCompilationTrackerState(
+                private InProgressState(
                     bool isFrozen,
                     Compilation compilationWithoutGeneratedDocuments,
                     CompilationTrackerGeneratorInfo generatorInfo,
-                    Compilation? staleCompilationWithGeneratedDocuments)
-                    : base(isFrozen, compilationWithoutGeneratedDocuments, generatorInfo)
+                    Compilation? staleCompilationWithGeneratedDocuments,
+                    ImmutableList<(ProjectState state, CompilationAndGeneratorDriverTranslationAction action)> pendingTranslationSteps)
+                    : base(isFrozen,
+                           compilationWithoutGeneratedDocuments,
+                           generatorInfo)
                 {
-                    // We're the non-final state.  As such, there is a strong correspondence between these two pieces of
-                    // state.  Specifically, if we're frozen, then the documents must be final.  After all, we do not
-                    // want to generate SG docs in the frozen state.  Conversely, if we're not frozen, the documents
-                    // must not be final.  We *must* generate the final docs from this state to get into the
-                    // FinalCompilationTrackerState.
-
-                    Contract.ThrowIfFalse(IsFrozen == generatorInfo.DocumentsAreFinal);
+                    // Note: Intermediate projects can be empty.
+                    Contract.ThrowIfTrue(pendingTranslationSteps is null);
 
+                    PendingTranslationSteps = pendingTranslationSteps;
                     StaleCompilationWithGeneratedDocuments = staleCompilationWithGeneratedDocuments;
                 }
-            }
-
-            /// <summary>
-            /// A state where we are holding onto a previously built compilation, and have a known set of transformations
-            /// that could get us to a more final state.
-            /// </summary>
-            private sealed class InProgressState : NonFinalWithCompilationTrackerState
-            {
-                /// <summary>
-                /// The list of changes that have happened since we last computed a compilation. The oldState corresponds to
-                /// the state of the project prior to the mutation.
-                /// </summary>
-                public ImmutableList<(ProjectState oldState, CompilationAndGeneratorDriverTranslationAction action)> IntermediateProjects { get; }
 
-                public InProgressState(
+                public static InProgressState Create(
```

Comment:
Consider getting rid of this method now? It's doing one line of interest and then calling the constructor...

Reply:
that line of interest is important though... if we get rid of this, someone could accidentally bypass it.

Reply:
I meant just move the line to the constructor -- then that way nobody can bypass it either.